### PR TITLE
Add ruggeduino support

### LINKED ISSUE #3: Ruggeduino Support
We need to support the ruggeduino. Needs #11 to work.

To do:

- [x] Add to HARDWARE_ENVIRONMENT (Console is blocked by https://github.com/j5api/j5/issues/621)
- [x] Add to Robot class
- [x] Add support for ignoring arduinos at init
- [x] Make list of ignored ruggeduinos available if present so that pyserial instances can be created
- [x] Consider adding pyserial as a dependency

e.g 

```python
from sr.robot3 import *

r = Robot(ignored_ruggeduinos=["SERIAL", "SERIAL2"])
```

### PR DESCRIPTION
**Pull request by trickeydan**
Theoretically closes #3

### CODE REVIEW

**Review by tpoomlmly — DISMISSED** [review]

**Comment by trickeydan** [comment]
@RealOrangeOne Requesting review from you specifically on the Python usage here.

Is it acceptable, especially around line 99. Not sure if it's too introspecty.

**Review by tpoomlmly — APPROVED** [review]

File: sr/robot3/robot.py [hunk 1]
```diff
@@ -71,6 +86,37 @@ def _init_auxilliary_boards(self) -> None:
...
+
+        self.ruggeduinos: BoardGroup[
+            Ruggeduino,
+            Backend,
+        ] = BoardGroup.get_board_group(
+            Ruggeduino,
+            IgnoredRuggeduinoBackend,
+        )
+
+        self.ignored_ruggeduinos = IGNORED
```
Comment (RealOrangeOne) [inline sr/robot3/robot.py]: Is there any reason to keep these hanging around?

File: sr/robot3/robot.py [hunk 2]
```diff
@@ -71,6 +86,37 @@ def _init_auxilliary_boards(self) -> None:
...
+        class IgnoredRuggeduinoBackend(ruggeduino_backend):  # type: ignore
+            """A backend that ignores some ruggeduinos."""
+
+            @classmethod
+            def is_arduino(cls, port: ListPortInfo) -> bool:
+                """Check if a ListPortInfo represents a valid Arduino derivative."""
+                if port.serial_number in self._ignored_ruggeduino_serials:
+                    IGNORED[port.serial_number] = port.device
+                    return False
+                return (port.vid, port.pid) in cls.USB_IDS
```
Comment (RealOrangeOne) [inline sr/robot3/robot.py]: Surely this checking should be part of the backend in itself already? Duplicating it here feels unnecessary, and the wrong place for it.

File: sr/robot3/robot.py [hunk 3]
```diff
@@ -71,6 +86,37 @@ def _init_auxilliary_boards(self) -> None:
         self.motor_boards = self._environment.get_board_group(MotorBoard)
         self.servo_boards = self._environment.get_board_group(ServoBoard)
 
+    def _init_ruggeduinos(self) -> None:
+        """
+        Initialise the ruggeduinos.
+
+        Ignore any that the user has specified.
+        """
+        IGNORED: Dict[str, str] = {}
```
Comment (RealOrangeOne) [inline sr/robot3/robot.py]: This shouldn't be upper case

**Review by RealOrangeOne — COMMENTED** [review]
I'm not seeing any introspection going on, although I do think the approach can probably be improved, or at the very least abstracted away

File: sr/robot3/robot.py (hunk 1 above)
Comment (trickeydan) [reply sr/robot3/robot.py]: Ignored ruggeduinos should be available for use by students with their own custom firmware.

Ignored = ignored by our code; students are free to go and spin up their own pyserial stuff.

We do want to expose the serial port paths to them though, which is why this is left lying around.

File: sr/robot3/robot.py (hunk 3 above)
Comment (trickeydan) [reply sr/robot3/robot.py]: 384c936

File: sr/robot3/robot.py (hunk 2 above)
Comment (trickeydan) [reply sr/robot3/robot.py]: 384c936

**Comment by trickeydan** [comment]
> I'm not seeing any introspection going on

```python
ruggeduino_backend = self._environment.get_backend(Ruggeduino)

class IgnoredRuggeduinoBackend(ruggeduino_backend):
```

This dynamic inheritance is the bit I'm concerned about. Requires type ignores for mypy, but also feels a bit dodgy. I can't think of a better way to do it though unfortunately.

**Pull request merged**